applets/wncklet/window-list.c

```diff
@@ -139,6 +146,166 @@ static void applet_change_background(MatePanelApplet* applet, MatePanelAppletBac
+
+	/* Resize window to fit thumbnail */
+	gtk_window_resize (GTK_WINDOW (tasklist->preview), width, height);
+
+	/* Set position at pointer, then re-adjust from there to just outside of the pointer */
+	gtk_window_set_position (GTK_WINDOW (tasklist->preview), GTK_WIN_POS_MOUSE);
+	gtk_window_get_position (GTK_WINDOW (tasklist->preview), &x_pos, &y_pos);
+
+	/* Get geometry of monitor where tasklist is located to calculate correct position of preview */
+	monitor = gdk_display_get_monitor_at_point (gdk_display_get_default (), x_pos, y_pos);
+	gdk_monitor_get_geometry (monitor, &monitor_geom);
+
+	/* Add padding to clear the panel */
+	switch (mate_panel_applet_get_orient (MATE_PANEL_APPLET (tasklist->applet)))
+	{
+		case MATE_PANEL_APPLET_ORIENT_LEFT:
+			x_pos = monitor_geom.width + monitor_geom.x - (width + tasklist->size) - PREVIEW_PADDING;
+			break;
+		case MATE_PANEL_APPLET_ORIENT_RIGHT:
+			x_pos = tasklist->size + PREVIEW_PADDING;
+			break;
+		case MATE_PANEL_APPLET_ORIENT_UP:
+			y_pos = monitor_geom.height + monitor_geom.y - (height + tasklist->size) - PREVIEW_PADDING;
+			break;
+		case MATE_PANEL_APPLET_ORIENT_DOWN:
+		default:
+			y_pos = tasklist->size + PREVIEW_PADDING;
+			break;
+	}
+
+	gtk_window_move (GTK_WINDOW (tasklist->preview), x_pos, y_pos);
+}
+
+static gboolean preview_window_draw (GtkWidget *widget, cairo_t *cr, GdkPixbuf *thumbnail)
+{
+	GtkStyleContext *context;
+
+	context = gtk_widget_get_style_context (widget);
+	gtk_render_icon (context, cr, thumbnail, 0, 0);
+
+	return FALSE;
+}
+
+static gboolean applet_enter_notify_event (WnckTasklist *tl, GList *wnck_windows, TasklistData *tasklist)
+{
+	GdkPixbuf *thumbnail;
+	WnckWindow *wnck_window = NULL;
+	int n_windows;
+
+	if (GTK_IS_WIDGET (tasklist->preview))
+		gtk_widget_destroy (tasklist->preview);
+
+	if (!tasklist->show_window_thumbnails || wnck_windows == NULL)
+		return FALSE;
+
+	n_windows = g_list_length (wnck_windows);
+	/* TODO: Display a list of stacked thumbnails for grouped windows. */
+	if (n_windows == 1)
+	{
+		GList* l = wnck_windows;
+		if (l != NULL)
+			wnck_window = (WnckWindow*)l->data;
+	}
+
+	if (wnck_window == NULL)
+		return FALSE;
+
+	thumbnail = preview_window_thumbnail (wnck_window, tasklist);
+
+	if (thumbnail == NULL)
+		return FALSE;
+
+	/* Create window to display preview */
+	tasklist->preview = gtk_window_new (GTK_WINDOW_POPUP);
+
+	gtk_widget_set_app_paintable (tasklist->preview, TRUE);
+	gtk_window_set_resizable (GTK_WINDOW (tasklist->preview), TRUE);
+
+	preview_window_reposition (tasklist, thumbnail);
+
+	gtk_widget_show (tasklist->preview);
+
+	g_signal_connect_data (G_OBJECT (tasklist->preview), "draw", G_CALLBACK (preview_window_draw), thumbnail, (GClosureNotify) g_object_unref, 0);
+
+	return FALSE;
+}
+
+static gboolean applet_leave_notify_event (WnckTasklist *tl, GList *wnck_windows, TasklistData *tasklist)
+{
+	g_return_val_if_fail (GTK_IS_WIDGET (tasklist->preview), FALSE);
```

Comment:
This can be optimized out...

Reply:
Since preview widget might not be created in applet_enter_notify_event you should just check if preview is not null before trying to destroy it.